internal/factsengine/gatherers/dir_scan.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	DirScanScanningError = entities.FactGatheringError{
+		Type:    "dir-scan-scanning-error",
+		Message: "error during directory scanning",
+	}
+)
+
+type DirScanDetails struct {
+	Name  string           `json:"name,omitempty"`
+	Owner string           `json:"owner"`
+	Group string           `json:"group"`
+	Files []DirScanDetails `json:"files,omitempty"`
+}
+
+type DirScanStatInfo struct {
+	Group string
+	Owner string
+}
+
+type DirScanResult map[string]DirScanDetails
+
+//go:generate mockery --name=UserSearcher
+type UserSearcher interface {
+	GetUsernameByID(userID string) (string, error)
+}
+
+//go:generate mockery --name=GroupSearcher
+type GroupSearcher interface {
+	GetGroupByID(groupID string) (string, error)
+}
+
+type DirScanGatherer struct {
+	fs            afero.Fs
+	userSearcher  UserSearcher
+	groupSearcher GroupSearcher
+}
+
+func NewDirScanGatherer(fs afero.Fs, userSearcher UserSearcher, groupSearcher GroupSearcher) *DirScanGatherer {
+	return &DirScanGatherer{fs: fs, userSearcher: userSearcher, groupSearcher: groupSearcher}
+}
+
+func NewDefaultDirScanGatherer() *DirScanGatherer {
+	cf := CredentialsFetcher{}
+	return NewDirScanGatherer(afero.NewOsFs(), &cf, &cf)
+}
+
+func (d *DirScanGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	log.Infof("Starting %s facts gathering process", DirScanGathererName)
+	facts := []entities.Fact{}
+
+	for _, requestedFact := range factsRequests {
+		if requestedFact.Argument == "" {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, &DirScanMissingArgumentError))
+			continue
+		}
+		scanResult, err := d.extractDirScanDetails(requestedFact.Argument)
+		if err != nil {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, DirScanScanningError.Wrap(err.Error())))
+			continue
+		}
+		factValue, err := mapDirScanResultToFactValue(scanResult)
+		if err != nil {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, DirScanScanningError.Wrap(err.Error())))
+			continue
+		}
+		facts = append(facts, entities.NewFactGatheredWithRequest(requestedFact, factValue))
+	}
+
+	log.Infof("Requested %s facts gathered", DirScanGathererName)
+
+	return facts, nil
+}
+
+func (d *DirScanGatherer) extractDirScanDetails(dirscanPath string) (DirScanResult, error) {
+	result := DirScanResult{}
+
+	matches, err := afero.Glob(d.fs, dirscanPath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, match := range matches {
+		resultKey := getDirScanResultKeyFromPath(match)
+
+		statInfo, err := d.getStatInfoForPath(match)
+		if err != nil {
+			return nil, err
+		}
+
+		if resultEntry, found := result[resultKey]; found {
+			resultEntry.Files = append(resultEntry.Files, DirScanDetails{
+				Name:  match,
+				Owner: statInfo.Owner,
+				Group: statInfo.Group,
+				Files: nil,
+			})
+			result[resultKey] = resultEntry
+			continue
+		}
+
+		newEntry := DirScanDetails{
+			Owner: statInfo.Owner,
```

Comment:
In fact, after testing in a real env, this approach looks wrong.
Here, we are creating a new "parent" folder with owner/group of the child. So the owner/group is inherited from the first file that matches the glob. Meaning that the parent folder doesn't have its own stats.
Anyway, this wouldn't have any impact if we decide to remove this "parent" folder entry